src/thor/route_matcher.cc

```diff
@@ -235,73 +237,85 @@ bool expand_from_node(const sif::mode_costing_t& mode_costing,
   return false;
 }
 
+valhalla::baldr::TimeInfo init_time_info(valhalla::baldr::GraphReader& reader,
+                                         valhalla::Options& options,
+                                         valhalla::baldr::DateTime::tz_sys_info_cache_t* tz_cache) {
+  graph_tile_ptr tile = nullptr;
+  const DirectedEdge* directededge = nullptr;
+  const NodeInfo* nodeinfo = nullptr;
+
+  // We support either the epoch timestamp that came with the trace point or
+  // a local date time which we convert to epoch by finding the first timezone
+  auto time_info = TimeInfo::invalid();
+  for (const auto& e : options.locations(0).path_edges()) {
+    GraphId graphid(e.graph_id());
+    if (!graphid.Is_Valid() || !reader.GetGraphTile(graphid, tile))
+      continue;
+    directededge = tile->directededge(graphid);
+    if (reader.GetGraphTile(directededge->endnode(), tile)) {
+      // get the timezone
+      nodeinfo = tile->node(directededge->endnode());
+      const auto* tz = DateTime::get_tz_db().from_index(nodeinfo->timezone());
+      if (!tz)
+        continue;
+      // if its timestamp based we need to convert that to a date time string on the location
+      if (!options.shape(0).has_date_time() && options.shape(0).time() != -1.0) {
+        options.mutable_shape(0)->set_date_time(
+            DateTime::seconds_to_date(options.shape(0).time(), tz, false));
+      }
+      return TimeInfo::make(*options.mutable_shape(0), reader, tz_cache, nodeinfo->timezone());
+    }
+  }
+  return TimeInfo::invalid();
+}
+
 } // namespace
 
+namespace valhalla {
+namespace thor {
+
 // For the route path using an edge walking method.
 bool RouteMatcher::FormPath(const sif::mode_costing_t& mode_costing,
                             const sif::TravelMode& mode,
-                            GraphReader& reader,
-                            const std::vector<meili::Measurement>& shape,
+                            baldr::GraphReader& reader,
                             valhalla::Options& options,
-                            std::vector<PathInfo>& path_infos) {
+                            std::vector<std::vector<PathInfo>>& legs) {
```

Comment:
i started refactoring this method to support multileg routes but there needs to be a little bit more refactoring in the guts of these algorithm to make it possible. i started trying it but it was getting too hair so i aborted that for now